File: algorithms/M/Maximum Candies Allocated to K Children/Maximum Candies Allocated to K Children.cpp
class Solution {
public:
    bool canSplit(vector<int>& candies, long long k, long long mid) {
        long long split = 0;
        for(int i = 0; i < candies.size(); ++i) {
            split += candies[i]/mid;
        }   
        if(split >= k)
            return true;
        else
            return false;
    }
    
    int maximumCandies(vector<int>& candies, long long k) {
        long long sum = 0;
        for(int i = 0; i < candies.size(); ++i) {
            sum += candies[i];
        }
        long long start = 1, end = sum/k;
        long long ans = 0;
        while(start <= end) {
            long long mid = (start + end)/2;
            if(canSplit(candies, k, mid)) {
                ans = mid;
                start = mid + 1;
            } else {
                end = mid-1;
            }
        }
        return ans;
    }
};

File: scripts/algorithms/M/Maximum Candies Allocated to K Children/Maximum Candies Allocated to K Children.py
# Runtime: 3294 ms (Top 14.29%) | Memory: 27.4 MB (Top 62.05%)
def canSplit(candies, mid, k):
    split = 0
    for i in candies:
        split += i//mid
    if split >= k:
        return True
    else:
        return False

class Solution:
    def maximumCandies(self, candies: List[int], k: int) -> int:
        end = sum(candies)//k
        start = 1
        ans = 0
        while start <= end:
            mid = (start + end)//2
            if canSplit(candies, mid, k):
                start = mid + 1
                ans = mid
            else:
                end = mid - 1
        return ans
